File: backend/app/schemas/voice_intake.py
"""Pydantic models for Phase 2 Step 2 - voice intake (see docs/phase2-step2-voice-intake.md §3)."""

from datetime import date
from typing import Any, Literal

from pydantic import BaseModel, Field


# Supported voice intents. Keep aligned with §5 of the design doc and
# services.voice_intake._DISPATCH (both referenced by string literal).
VoiceIntent = Literal[
    "deworming",
    "vaccination",
    "weight",
    "routine",
    "unknown",
]


class VoiceIntakeDraft(BaseModel):
    """Structured, normalised draft produced by STT+LLM+normalisation.

    Every field is optional because missing ones go into `missing_fields`
    and the front-end will prompt for them. We deliberately don't
    sub-type per intent so the response stays a single stable shape; the
    server validates shape-vs-intent when the user confirms.
    """

    pet_id: int | None = None
    pet_name: str | None = None
    note: str | None = None

    # deworming
    deworming_type: Literal["internal", "external", "combined"] | None = None
    dewormed_at: date | None = None

    # vaccination
    vaccine_name: str | None = None
    vaccinated_at: date | None = None

    # weight
    weight_kg: float | None = None
    weighed_at: date | None = None

    # routine
    routine_type: Literal["bath", "nail_trim", "grooming"] | None = None
    routine_at: date | None = None


class VoiceIntakeResponse(BaseModel):
    """Returned by `POST /api/v1/voice/intake`."""

    request_id: str
    status: Literal[
        "stt_failed",
        "intent_unknown",
        "draft_pending",
    ]
    transcript: str | None = None
    intent: VoiceIntent | None = None
    confidence: int | None = None
    needs_confirm: bool = True
    draft: VoiceIntakeDraft | None = None
    missing_fields: list[str] = Field(default_factory=list)


class VoiceIntakeConfirmRequest(BaseModel):
    """Payload for `POST /api/v1/voice/intake/confirm`.

    `payload` is a free-form dict so each intent can pass its own
    field set; the server re-validates through the intent-specific
    Pydantic schema (DewormingCreate, etc.) before calling the
    existing write services.
    """

    request_id: str
    intent: VoiceIntent
    payload: dict[str, Any]


class VoiceIntakeConfirmResponse(BaseModel):
    request_id: str
    status: Literal["confirmed"]
    entity_type: str
    entity_id: int
    entity: dict[str, Any]

File: frontend/lib/models/voice_intake.dart
// Phase 2 Step 2 — voice intake models (docs/phase2-step2-voice-intake.md §3).

enum VoiceIntakeStatus {
  sttFailed,
  intentUnknown,
  draftPending,
}

VoiceIntakeStatus _statusFromString(String s) {
  switch (s) {
    case 'stt_failed':
      return VoiceIntakeStatus.sttFailed;
    case 'intent_unknown':
      return VoiceIntakeStatus.intentUnknown;
    case 'draft_pending':
      return VoiceIntakeStatus.draftPending;
  }
  throw ArgumentError('Unknown voice intake status: $s');
}

enum VoiceIntent {
  deworming,
  vaccination,
  weight,
  routine,
  unknown,
}

VoiceIntent? voiceIntentFromString(String? s) {
  if (s == null) return null;
  switch (s) {
    case 'deworming':
      return VoiceIntent.deworming;
    case 'vaccination':
      return VoiceIntent.vaccination;
    case 'weight':
      return VoiceIntent.weight;
    case 'routine':
      return VoiceIntent.routine;
    case 'unknown':
      return VoiceIntent.unknown;
  }
  return null;
}

String voiceIntentApiValue(VoiceIntent intent) {
  switch (intent) {
    case VoiceIntent.deworming:
      return 'deworming';
    case VoiceIntent.vaccination:
      return 'vaccination';
    case VoiceIntent.weight:
      return 'weight';
    case VoiceIntent.routine:
      return 'routine';
    case VoiceIntent.unknown:
      return 'unknown';
  }
}

String voiceIntentLabel(VoiceIntent intent) {
  switch (intent) {
    case VoiceIntent.deworming:
      return '驱虫';
    case VoiceIntent.vaccination:
      return '疫苗';
    case VoiceIntent.weight:
      return '体重';
    case VoiceIntent.routine:
      return '日常';
    case VoiceIntent.unknown:
      return '未知';
  }
}

/// A structured draft built by the backend from the raw STT/LLM output.
///
/// Every field is optional — missing fields are listed in
/// [VoiceIntakeResponse.missingFields] and the UI prompts for them
/// before calling `confirm`.
class VoiceIntakeDraft {
  final int? petId;
  final String? petName;
  final String? note;

  // deworming
  final String? dewormingType; // internal | external | combined
  final String? dewormedAt;    // YYYY-MM-DD

  // vaccination
  final String? vaccineName;
  final String? vaccinatedAt;

  // weight
  final double? weightKg;
  final String? weighedAt;

  // routine
  final String? routineType; // bath | nail_trim | grooming
  final String? routineAt;

  const VoiceIntakeDraft({
    this.petId,
    this.petName,
    this.note,
    this.dewormingType,
    this.dewormedAt,
    this.vaccineName,
    this.vaccinatedAt,
    this.weightKg,
    this.weighedAt,
    this.routineType,
    this.routineAt,
  });

  factory VoiceIntakeDraft.fromJson(Map<String, dynamic> json) {
    final rawWeight = json['weight_kg'];
    double? weight;
    if (rawWeight is num) {
      weight = rawWeight.toDouble();
    } else if (rawWeight is String) {
      weight = double.tryParse(rawWeight);
    }
    return VoiceIntakeDraft(
      petId: json['pet_id'] as int?,
      petName: json['pet_name'] as String?,
      note: json['note'] as String?,
      dewormingType: json['deworming_type'] as String?,
      dewormedAt: json['dewormed_at'] as String?,
      vaccineName: json['vaccine_name'] as String?,
      vaccinatedAt: json['vaccinated_at'] as String?,
      weightKg: weight,
      weighedAt: json['weighed_at'] as String?,
      routineType: json['routine_type'] as String?,
      routineAt: json['routine_at'] as String?,
    );
  }
}

class VoiceIntakeResponse {
  final String requestId;
  final VoiceIntakeStatus status;
  final String? transcript;
  final VoiceIntent? intent;
  final int? confidence;
  final bool needsConfirm;
  final VoiceIntakeDraft? draft;
  final List<String> missingFields;

  const VoiceIntakeResponse({
    required this.requestId,
    required this.status,
    required this.transcript,
    required this.intent,
    required this.confidence,
    required this.needsConfirm,
    required this.draft,
    required this.missingFields,
  });

  factory VoiceIntakeResponse.fromJson(Map<String, dynamic> json) {
    return VoiceIntakeResponse(
      requestId: json['request_id'] as String,
      status: _statusFromString(json['status'] as String),
      transcript: json['transcript'] as String?,
      intent: voiceIntentFromString(json['intent'] as String?),
      confidence: json['confidence'] as int?,
      needsConfirm: (json['needs_confirm'] as bool?) ?? true,
      draft: json['draft'] == null
          ? null
          : VoiceIntakeDraft.fromJson(json['draft'] as Map<String, dynamic>),
      missingFields: (json['missing_fields'] as List<dynamic>? ?? [])
          .map((e) => e as String)
          .toList(),
    );
  }
}

class VoiceIntakeConfirmResult {
  final String requestId;
  final String entityType;
  final int entityId;
  final Map<String, dynamic> entity;

  const VoiceIntakeConfirmResult({
    required this.requestId,
    required this.entityType,
    required this.entityId,
    required this.entity,
  });

  factory VoiceIntakeConfirmResult.fromJson(Map<String, dynamic> json) {
    return VoiceIntakeConfirmResult(
      requestId: json['request_id'] as String,
      entityType: json['entity_type'] as String,
      entityId: json['entity_id'] as int,
      entity: (json['entity'] as Map).cast<String, dynamic>(),
    );
  }
}
